### DISCUSSION TITLE
Benchmarking the Performance (FPS) of Different Stages of the Pipeline

### DISCUSSION THREAD

**Discussion by vaghawan** [miguelaeh/pipeless#149] [Q&A]
Hi, 

Thanks a million for such a wonderful and seamless framework. If you've to deal with tools e.g deepstream and such, you know how thankful one can be of pipeless. 

Now to my problem, I need to benchmark the performance at the each stage of the pipeline, to begin with I need to capture the average frame per second (FPS) performance of each stage. I've 6 stages, and my goal is to stream back the results in almost real time, but to even begin to assess it, I need a way to capture the FPS first. 

Can somebody who may have the idea or who may have already done it give me some input, or share how can I possibly do it? 

Thank you in advance.

**Comment by miguelaeh** [comment]
Hi @vaghawan ,
A way in which you can measure that is using the KV store to store the values you canculate on the stage. For example, you can calculate how long the code in your stage takes to run and then store the value or add it to a mean or whatever you need. Do you think that will help you obtain what you need?

**Comment by vaghawan** [comment]
Hi @miguelaeh , Thank you! 

I'm referring to this page : https://www.pipeless.ai/docs/docs/v1/kvstore 

I believe `pipeless_kvs_set("my_key", "My value")` , this is to be set only once, but I need to measure the time it took in the process in every single call of the hook to determine the average processing time. 

```
   def hook(frame_data, _):
    start_time = time.time()
    frame = frame_data["original"].view()
    yolo_input_shape = (640, 640, 3) # h,w,c
    frame = resize_with_padding(frame, yolo_input_shape)
    frame = np.array(frame) / 255.0 # Normalize pixel values
    frame = np.transpose(frame, axes=(2,0,1)) # Convert to c,h,w
    inference_inputs = frame.astype("float32")
    frame_data['inference_input'] = inference_inputs
    end_time = time.time()
    # Write the time difference to the file
    with open(log_file, "a") as f:
        f.write(f"Time taken: {(end_time - start_time) * 1000} milliseconds\n")
``` 

Say if I were to use the KV, would I be setting the value in each hook call, wouldn't that replace the old value? 

Thanks

**Reply by miguelaeh** [reply]
Hi @vaghawan , 
That's right. The KV store will override previous values. You can try to prepend the frame number to the key, which is part of the frame data in order to identify the frames.
Also, how do you expect to get the FPS data? Do you want to just ser it or do you need to export it somewhere else? I think both things can be done directly in that Python code

**Reply by vaghawan** [reply]
Hi @miguelaeh , I need to store it, yes, I figured that I could use frame_id as a key, but I'm not sure how to retrieve it or write it somewhere after the stage is finished. I need to re-analyze the data to calculate average etc. 

For now, it's very simple, I don't know if we can even refer to this as FPS, but like in the code above, I'm going to capture how much time it took to execute the pre-process, inference (process) and post-process, then sum then up to see how long it takes in total for a single frame. 

I'm not sure how you guys execute the benchmark presented in this page? https://www.pipeless.ai/docs/docs/v1/benchmark 

If you could share that, I'm right now working with Jetson Orien, I could help adding the benchmark.

Also, since I've worked on something that needs multiple stages, and usage of custom data and kv, I could actually write the example which we can probably include here, I don't know. 

Thanks
V

**Reply by miguelaeh** [reply]
> I'm not sure how you guys execute the benchmark presented in this page? https://www.pipeless.ai/docs/docs/v1/benchmark

Those were created by exploring the metrics. There is an input timestamp on the frame data too, which indicates the time at which the frame is received. The difference between that and the last hook you execute is the whole time the frame is on the pipeline (assuming no output video).

> I could actually write the example which we can probably include here, I don't know.

That would be awesome. I have tested it out in a Jetson Xavier so far. There is a lower performance on Jetson yet since their CPUs are very basic and it still needs to perform a copy of the frames internally. This is something we have pending to address.

**Reply by vaghawan** [reply]
Hi @miguelaeh , Thanks!! So, how do you access the timstamps in the last hook? Is there any inherent timestamps at the end like `input_ts` ? ( https://www.pipeless.ai/docs/docs/v1/getting-started ) 

> That would be awesome. I have tested it out in a Jetson Xavier so far. There is a lower performance on Jetson yet since their CPUs are very basic and it still needs to perform a copy of the frames internally. This is something we have pending to address.


I can  help with the benchmark on Jetson Orien, I saw that using TensorRT models give the best performance. For usual model, yes it's very slow. 

Thanks

**Reply by vaghawan** [reply]
Hi @miguelaeh , I've the average benchmark for some models on Jetson Orien device. How do you want me to forward it to you? I just have average and standard deviation of the inference time for Yolo TensorRT converted models and some other models, which may not so relevant for general usage.

**Reply by miguelaeh** [reply]
Hi @vaghawan ,
Regarding the timestamps you can just use the time method on the hook. That's the moment when it gets executed.

Thank you very much for the Orien benchmark!! You can open a PR on the [docs repo](https://github.com/pipeless-ai/docs) directly to add them, I will be more than hsppy to review it

**Reply by vaghawan** [reply]
Thanks @miguelaeh , I'm going to do that very soon, I already have the numbers. By the way, one thing, is it possible to dump the kv storage (user data) when the whole pipeline is finished? I could record the individual timestamps that way too. 
It appears writing the time for each frame by appending to a txt file seems to add latency in itself. I would like to export the frame_ids at the end too, so that I can keep track of them in some kind of database.

**Reply by miguelaeh** [reply]
I guess we could implement such dump on the core itself. Everything you add to the pipeline will add latency, so I guess the tolerance depends on the use case

**Reply by vaghawan** [reply]
Hi @miguelaeh , Thanks, sent in the PR for the benchmark. 

> I guess we could implement such dump on the core itself.

I think this will be useful.
